include/slate/Tile.hh

```diff
@@ -1123,6 +1124,67 @@ void Tile<scalar_t>::recv(int src, MPI_Comm mpi_comm, Layout layout, int tag)
     // by receiving less / compacted data
 }
 
+//------------------------------------------------------------------------------
+/// Receives tile from MPI rank src using immediate mode
+///
+/// @param[in] src
+///     Source MPI rank in mpi_comm.
+///
+/// @param[in] mpi_comm
+///     MPI communicator.
+///
+/// @param[in] layout
+///     Indicates the Layout (ColMajor/RowMajor) of the received data.
+///     WARNING: need to call tileLayout(...) to properly set the layout of the
+///              origin matrix tile afterwards.
+/// @param[in] tag
+///     MPI tag
+///
+/// @param[out] request
+///     MPI request object
+///
+// todo need to copy or verify metadata (sizes, op, uplo, ...)
+template <typename scalar_t>
+void Tile<scalar_t>::irecv(int src, MPI_Comm mpi_comm, Layout layout,
+                           int tag, MPI_Request* request)
+{
+    trace::Block trace_block("MPI_Recv");
+
+    // If no stride.
+    if (this->isContiguous()) {
+        // Use simple recv.
+        int count = mb_*nb_;
+
+        slate_mpi_call(
+            MPI_Irecv(data_, count, mpi_type<scalar_t>::value, src, tag,
+                     mpi_comm, request));
+    }
+    else {
+        // Otherwise, use strided recv.
+        int count = layout_ == Layout::ColMajor ? nb_ : mb_;
+        int blocklength = layout_ == Layout::ColMajor ? mb_ : nb_;
+        int stride = stride_;
+        MPI_Datatype newtype;
+
+        slate_mpi_call(
+            MPI_Type_vector(
+                count, blocklength, stride, mpi_type<scalar_t>::value,
+                &newtype));
+
+        slate_mpi_call(MPI_Type_commit(&newtype));
```

Comment:
Side note: A question that I have always wanted to get an answer: is it okay to commit /free over and over the MPI_Type?

Reply:
I've always assumed that it's safe, but that there might be a performance overhead.  But, I haven't ever looked into it either.

Someone could probably implement a cache since I think there's only be 4 distinct datatypes for, e.g., a ScaLAPACK matrix.